src/rust/ide/src/ide.rs

```diff
@@ -0,0 +1,184 @@
+        let client_json   = language_server::Client::new(json_ws);
+        let client_binary = binary::Client::new(logger,binary_ws);
+        crate::executor::global::spawn(client_json.runner());
+        crate::executor::global::spawn(client_binary.runner());
+        let connection_json   = language_server::Connection::new(client_json,client_id).await?;
+        let connection_binary = binary::Connection::new(client_binary,client_id).await?;
+        Ok(controller::Project::new(logger,connection_json,connection_binary,project_name))
+    }
+
+    /// Creates a new project and returns its metadata, so the newly connected project can be
+    /// opened.
+    pub async fn create_project
+    ( logger          : &Logger
+    , project_manager : &impl project_manager::API
+    , name            : &String) -> FallibleResult<ProjectMetadata> {
+        info!(logger, "Creating a new project named `{name}`.");
+        let id = project_manager.create_project(name).await?.project_id;
+        Ok(ProjectMetadata {
+            id,
+            name        : ProjectName {name:name.clone()},
+            last_opened : None,
+        })
+    }
+
+    /// Open the named project or create a new project if it doesn't exist.
+    pub async fn open_project_or_create_new
+    ( logger : &Logger
+    , project_manager : &impl project_manager::API
+    , project_name    : &String) -> FallibleResult<controller::Project> {
+        let projects_to_list = constants::MAXIMUM_LISTABLE_PROJECTS;
+        let response     = project_manager.list_recent_projects(&projects_to_list).await?;
+        let mut projects = response.projects.iter();
+        let project      = projects.find(|project_metadata| {
+            project_metadata.name.name == *project_name
+        });
+        let project_metadata = if let Some(project) = project {
+            project.clone()
+        } else {
+            println!("Attempting to create {}", project_name);
+            Self::create_project(logger,project_manager,project_name).await?
+        };
+        let endpoints = project_manager.open_project(&project_metadata.id).await?;
+        Self::open_project(logger,endpoints.language_server_json_address,
+            endpoints.language_server_binary_address,&project_metadata.name.name).await
+    }
+
+    /// Open most recent project or create a new project if none exists.
+    pub async fn open_most_recent_project_or_create_new
+    ( logger          : &Logger
+    , project_manager : &impl project_manager::API) -> FallibleResult<controller::Project> {
+        let projects_to_list = 1;
+        let mut response     = project_manager.list_recent_projects(&projects_to_list).await?;
+        let project_metadata = if let Some(project) = response.projects.pop() {
+            project
+        } else {
+            let project_name = constants::DEFAULT_PROJECT_NAME.to_string();
+            Self::create_project(logger,project_manager,&project_name).await?
```

Comment:
I completely do not understand this logic. It takes `&str`, converts it to `String` only to convert it back to `&str` in the next line